### PR TITLE
fix: Handle invalid (zero) capacity of changelog and roots reliably

### PR DESCRIPTION
**Pull request by vadorovsky**
* Validate changelog and roots size not only in `new()` method, but also in `init()` and all wrapper types (zero-copy and copy).
* Add SBF tests which initialize trees with zero capacities and expect an error.
* Add SBF tests with changelog and root capacity 1. They are expected to succeed.

### CODE REVIEW

File: test-programs/account-compression-test/tests/merkle_tree_tests.rs [hunk 1]
```diff
@@ -1614,6 +1615,11 @@ pub async fn fail_initialize_state_merkle_tree_and_nullifier_queue_invalid_sizes
 /// parameters:
 ///
 /// 1. Merkle tree height (different than 26).
+/// 2. Merkle tree canopy depth (different than 10).
+/// 3. Merkle tree changelog size (zero).
+/// 4. Merkle tree roots size (zero).
+/// 5. Merkle tree close threshold (any).
+/// 6. Queue sequence threshold (lower than roots + safety margin).
```
Comment (ananas-block) [inline test-programs/account-compression-test/tests/merkle_tree_tests.rs:1622]: Which lines are tests for 4,5,6?

**Review by ananas-block — APPROVED** [review]

**Review by ananas-block — CHANGES_REQUESTED** [review]
Tests are too long we need to reduce the test iterations.
previous time was 23 not its 35 min

File: test-programs/account-compression-test/tests/address_merkle_tree_tests.rs [hunk 2]
```diff
@@ -915,29 +966,29 @@ async fn update_address_merkle_tree_failing_tests_default() {
...
-                        sequence_threshold: roots_size + SAFETY_MARGIN,
-                        network_fee: None,
-                    },
-                )
-                .await;
+    for changelog_size in (1..=5000).step_by(1000) {
+        for roots_size in (changelog_size..=5000).step_by(1000) {
+            for queue_capacity in [5003, 6857, 7901] {
+                for address_changelog_size in (250..1000).step_by(250) {
+                    update_address_merkle_tree_failing_tests(
```
Comment (ananas-block) [inline test-programs/account-compression-test/tests/address_merkle_tree_tests.rs:973]: do these failing tests depend on the different queue sizes etc?
if not lets remove these iterations

File: test-programs/account-compression-test/tests/merkle_tree_tests.rs [hunk 3]
```diff
@@ -192,26 +192,27 @@ async fn test_init_and_insert_into_nullifier_queue_default() {
...
-                    capacity: queue_capacity,
-                    sequence_threshold: roots_size + SAFETY_MARGIN,
-                    network_fee: None,
-                },
-            )
-            .await;
+    for changelog_size in (1..=5000).step_by(1000) {
+        for roots_size in (changelog_size..=5000).step_by(1000) {
+            for queue_capacity in [5003, 6857, 7901] {
+                test_init_and_insert_into_nullifier_queue(
```
Comment (ananas-block) [inline test-programs/account-compression-test/tests/merkle_tree_tests.rs:198]: I think this one makes sense to iterate over multiple capacities.
I don't know how the underlying test works but it's probably enough to just do a couple inserts and empty the queue no need to do a lot of transactions here.

File: test-programs/account-compression-test/tests/address_merkle_tree_tests.rs [hunk 4]
```diff
@@ -1288,29 +1339,29 @@ async fn test_address_merkle_tree_and_queue_rollover_default() {
...
+                for address_changelog_size in (250..1000).step_by(250) {
+                    address_merkle_tree_and_queue_rollover(
+                        &AddressMerkleTreeConfig {
+                            height: ADDRESS_MERKLE_TREE_HEIGHT as u32,
+                            changelog_size,
+                            roots_size,
+                            canopy_depth: ADDRESS_MERKLE_TREE_CANOPY_DEPTH,
+                            address_changelog_size,
+                            network_fee: Some(5000),
+                            rollover_threshold: Some(95),
```
Comment (ananas-block) [inline test-programs/account-compression-test/tests/address_merkle_tree_tests.rs:1354]: Iterations make sense!
We can just put the rollover threshold to zero to test this one, that should be faster.
Imo we should still keep one test that advances the slot and modifies the account (idk whether these replaced that original test which did that.)

**Review by ananas-block — APPROVED** [review]
